proposed/0000-crust-refactor/crust-library.md

```diff
@@ -0,0 +1,250 @@
+object, that can be disassembled into components to be used independently.
+
+```rust
+struct Service {
+    /// Used to accept incoming connections.
+    pub acceptor: ServiceListener,
+    /// Used to control the state of the beacon.
+    pub beacon_state: ServiceBeaconState,
+    /// Used to receive endpoints broadcasted by peers on the local network.
+    pub beacon_receiver: ServiceBeaconReceiver,
+    /// Used to control the service.
+    pub controller: ServiceController,
+}
+```
+
+The user moves `acceptor` to where they want to wait for incoming connections
+and `beacon_receiver` to where they want to wait for incoming beacon messages.
+
+`controller` is used to perform almost all the other functionality of crust. it
+implements `Sync` and it's methods borrow `self` immutably allowing it to be
+borrowed and used throughout the rest of the program. `controller` can be used to
+ * Add/remove/list listening endpoints on the service's internal `Acceptor`.
+ * Tell the service about external hole-punching servers.
+ * Get the local hole-punching server's addresses.
+ * Perform connections and rendezvous connections.
+ * Read the cache of recorded endpoints.
+
+```rust
+/// Drop this type to shutdown the service.
+impl ServiceController {
+    /// Accepting
+    pub fn add_listener(&self, addr: ListenEndpoint);
+    pub fn remove_listener(&self, addr: ListenEndpoint);
+    pub fn accepting_endpoints<'c>(&'c self) -> AcceptingEndpoints<'c>
+
+    /// Hole punching
+    pub fn mapping_context(&self) -> &MappingContext;
+    pub fn add_hole_puncher_server(&self, server: HolePunchServerAddr);
+    pub fn hole_punch_addresses(&self, bop_handle: &BopHandle)
+        -> BopResult<Vec<HolePunchServerAddr>>
+
+    /// Get a mapped udp socket using the `Service's internal `MappingContext`
+    pub fn mapped_udp_socket<'c>(&'c self, bop_handle: &BopHandle)
+        -> BopResult<MappedUdpSocket>;
+    /// Connect a `MappedUdpSocket`.
+    pub fn utp_rendezvous_connect(&self, bop_handle: &BopHandle,
+                                         mapped_socket: MappedUdpSocket,
+                                         their_info: UdpRendezvousInfo)
+        -> BopResult<Stream>
+
+    /// Connecting
+    pub fn connect(&self, bop_handle: &BopHandle, endpoint: Endpoint)
+        -> BopResult<Stream>
+
+    /// Cacheing/Bootstrapping
+    pub fn cache_endpoint(&self, endpoint: Endpoint)
+    pub fn iter_endpoint_cache<'c>(&'c self) -> EndpointCacheIterator<'c>
+}
+```
+
+The service's beacon state (enabled or disabled) is controlled via a separate
+object. This is just to take advantage of the fact that - being a simple
+boolean - it's state can be represented at the type level. The user can enforce
+compile-time guarantees about the state of the beacon during particular parts
+of their code, or they can just stick the `ServiceBeaconState` in a `Mutex` and
+use it the same way they use the `ServiceController`.
```

Comment:
It seems a bit inconsistent that `ServiceBeaconState` is treated like an immutable boolean value with respect to enabled/disabled, but provides a setter for the period. E. g. implementing something like `fn toggle(state: &mut ServiceBeaconState)` is impossible, whereas writing `fn adjust_period(state: &mut ServiceBeaconState)` is straightforward.

And I think putting the state in a `Mutex` is pretty cumbersome (but maybe I'm doing it wrong): To move it there you have to disassemble the `Service` first (it can't keep ownership of the object), and I think e. g. enabling the beacon is only really feasible with a `Mutex<Option<ServiceBeaconState>>` because before calling `enable`, you need to move the object out of the `Mutex`:
```rust
let mut guard = state_mutex.lock().unwrap();
if let Some(state) = mem::replace(&mut *guard, None) {
    *guard = Some(state.enable().into());
}
```

Probably the easiest solution is to just offer both:
```rust
    fn enable(&mut self);
    fn to_enabled(self) -> ServiceBeaconStateEnabled;
```